### DISCUSSION TITLE
Building component resources in Go

### DISCUSSION THREAD

**Discussion by daenney** [pulumi/pulumi#11694] [Pulumi in Practice]
I've started to take a look at Component Resources and as a concept they make sense to me, but the documentation of how to (g|G)o about them has left me a bit confused. I'm going based on this: https://www.pulumi.com/docs/intro/concepts/resources/components/ and using the Go programming language.

There's a couple of things I'm struggling with:
* Based on the code examples, it seems all child resources of a Component Resource have to be created within the constructor method?
* The `RegisterOutputs` shows an example of adding a key named `bucketDnsName` with the value of `bucket.BucketDomainName`. What if I create 2 buckets?
* How am I supposed to figure out what the keys in the `pulumi.ProviderMap` should be? For example, why is it `kubernetes` and not `k8s`?
* There's not really an example of how to take inputs. How do I go about that? I assume there's some kind of `MyComponentArgs` struct I should make and not explode the constructor arguments?
* The [AWS Quickstart VPC](https://github.com/pulumi/pulumi-aws-quickstart-vpc) has something that [looks an awful lot like a component resource](https://github.com/pulumi/pulumi-aws-quickstart-vpc/blob/main/sdk/go/aws/vpc.go) but it also has tons of additional generated code. Is that related, do I care and how do I go about generating that? Its last release was in 2021 so I'm not sure if this still applies?

If the AWS Quickstart VPC isn't what I should be looking at, does anyone have a real world example of a component resource in Go in a public repo they would be willing to share?

**Comment by daenney** [comment]
I ended up finding https://github.com/pulumi/workshops/tree/09d39e18662769a3008d985bafa577e56219f009/building-reusable-abstractions/2-component-resources/go after doing some global GitHub search which explains a great deal.

I'm still somewhat confused about the `provider.Map()` keys in the previous example as well as the `RegisterResourceOutputs` and how/if that relates to `ctx.Export()`?

**Comment by daenney** [comment]
I ended up asking ChatGPT, which at least solved my question on the inputs :smile:.

> Can you teach me how to write a Component Resource for Pulumi in the Go programming language?

Sure! A Component Resource in Pulumi is a way to group related resources into a reusable unit, making it easier to manage and deploy your infrastructure. Here's an example of a simple Component Resource in Go:

```go
package main

import (
	"github.com/pulumi/pulumi-aws/sdk/v2/go/aws/ec2"
	"github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		_, err := NewVpc(ctx, "example-vpc", &VpcArgs{
			CidrBlock: pulumi.String("10.0.0.0/16"),
		})
		if err != nil {
			return err
		}

		return nil
	})
}

type VpcArgs struct {
	CidrBlock pulumi.StringInput
}

func NewVpc(ctx *pulumi.Context, name string, args *VpcArgs) (*ec2.Vpc, error) {
	vpc, err := ec2.NewVpc(ctx, name, &ec2.VpcArgs{
		CidrBlock: args.CidrBlock,
	})
	if err != nil {
		return nil, err
	}

	return vpc, nil
}
```
In this example, we have created a new Component Resource called NewVpc that creates an AWS VPC. The NewVpc function takes in a Pulumi context, a name for the resource, and an instance of VpcArgs which allows us to pass in the CIDR block for the VPC. The NewVpc function then creates an instance of an AWS VPC using the ec2.NewVpc function from the pulumi-aws library. The created VPC is returned from the function so it can be used elsewhere in your code.

To use this Component Resource, you would call the NewVpc function in your main code and pass in the necessary arguments. This allows you to abstract away the details of creating a VPC, making your code easier to read and maintain.

---

> In some examples I've seen, there's a ProviderMap that also used in component resources to inherit a resource provider from a parent resource. How do I know what the key should be named in the ProviderMap?

The key in the ProviderMap should be named based on the package name of the resource provider that you want to inherit from the parent resource.

For example, if you want to inherit the AWS provider from the parent resource, the key in the ProviderMap should be named "aws" because the AWS provider is defined in the "github.com/pulumi/pulumi-aws/sdk/go/aws" package.

**Comment by mohammadasim** [comment]
This is how I have created a component and it works.
```
package main

import (
	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/dynamodb"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

type DynamodbComponentArgs struct {
	tableBillingMode         pulumi.StringInput
	studentsTableName        pulumi.StringInput
	classesTableName         pulumi.StringInput
	attendanceTableName      pulumi.StringInput
	addressTableName         pulumi.StringInput
	studentParrentsTableName pulumi.StringInput
}

type dynamodbComponent struct {
	pulumi.ResourceState

	//Output properties of the component
	StudentTableID         pulumi.StringOutput `pulumi:"studentTableID"`
	ClassesTableID         pulumi.StringOutput `pulumi:"classesTableID"`
	AttendanceTableID      pulumi.StringOutput `pulumi:"attendanceTableID"`
	StudentParrentsTableID pulumi.StringOutput `pulumi:"studentParrentsTableID"`
	AddressTableID         pulumi.StringOutput `pulumi:"addressTableID"`
}

func NewDynamodbComponent(ctx *pulumi.Context, name string, args *DynamodbComponentArgs, opts ...pulumi.ResourceOption) (*dynamodbComponent, error) {
	dynamodbComponent := &dynamodbComponent{}
	err := ctx.RegisterComponentResource("custom:module:DynamodbComponent", name, dynamodbComponent, opts...)
	if err != nil {
		return nil, err
	}
	// Create the student Dynamodb Table
	studentsTable, err := dynamodb.NewTable(ctx, "students", &dynamodb.TableArgs{
		Name:        args.studentsTableName,
		BillingMode: args.tableBillingMode,
		HashKey:     pulumi.String("student_id"),
		Attributes: dynamodb.TableAttributeArray{
			dynamodb.TableAttributeArgs{
				Name: pulumi.String("student_id"),
				Type: pulumi.String("S"),
			},
		},
	},
		pulumi.Parent(dynamodbComponent),
	)
	if err != nil {
		return nil, err
	}
	// Create the class Dynamodb Table
	classesTable, err := dynamodb.NewTable(ctx, "classes", &dynamodb.TableArgs{
		Name:        args.classesTableName,
		BillingMode: args.tableBillingMode,
		HashKey:     pulumi.String("class_id"),
		RangeKey:    pulumi.String("year"),
		Attributes: dynamodb.TableAttributeArray{
			dynamodb.TableAttributeArgs{
				Name: pulumi.String("class_id"),
				Type: pulumi.String("S"),
			},
			dynamodb.TableAttributeArgs{
				Name: pulumi.String("year"),
				Type: pulumi.String("N"),
			},
		},
	},
		pulumi.Parent(dynamodbComponent),
	)
	if err != nil {
		return nil, err
	}
	// Create the attendance Dynamodb Table
	attendanceTable, err := dynamodb.NewTable(ctx, "attendance", &dynamodb.TableArgs{
		Name:        args.attendanceTableName,
		BillingMode: args.tableBillingMode,
		HashKey:     pulumi.String("student_id"),
		RangeKey:    pulumi.String("date"),
		Attributes: dynamodb.TableAttributeArray{
			dynamodb.TableAttributeArgs{
				Name: pulumi.String("student_id"),
				Type: pulumi.String("S"),
			},
			dynamodb.TableAttributeArgs{
				Name: pulumi.String("date"),
				Type: pulumi.String("S"),
			},
		},
	},
		pulumi.Parent(dynamodbComponent),
	)
	if err != nil {
		return nil, err
	}
	// Create the address Dynamodb Table
	addressTable, err := dynamodb.NewTable(ctx, "address", &dynamodb.TableArgs{
		Name:        args.addressTableName,
		BillingMode: args.tableBillingMode,
		HashKey:     pulumi.String("address_id"),
		Attributes: dynamodb.TableAttributeArray{
			dynamodb.TableAttributeArgs{
				Name: pulumi.String("address_id"),
				Type: pulumi.String("S"),
			},
		},
	},
		pulumi.Parent(dynamodbComponent),
	)
	if err != nil {
		return nil, err
	}
	// Create the studentParrents Dynamodb Table
	studentParrentsTable, err := dynamodb.NewTable(ctx, "studentParrents", &dynamodb.TableArgs{
		Name:        args.studentParrentsTableName,
		BillingMode: args.tableBillingMode,
		HashKey:     pulumi.String("parents_id"),
		Attributes: dynamodb.TableAttributeArray{
			dynamodb.TableAttributeArgs{
				Name: pulumi.String("parents_id"),
				Type: pulumi.String("S"),
			},
		},
	},
		pulumi.Parent(dynamodbComponent),
	)
	if err != nil {
		return nil, err
	}
	dynamodbComponent.StudentTableID = studentsTable.ID().ToStringOutput()
	dynamodbComponent.ClassesTableID = classesTable.ID().ToStringOutput()
	dynamodbComponent.AttendanceTableID = attendanceTable.ID().ToStringOutput()
	dynamodbComponent.AddressTableID = addressTable.ID().ToStringOutput()
	dynamodbComponent.StudentParrentsTableID = studentParrentsTable.ID().ToStringOutput()

	return dynamodbComponent, err
}
```
This is how I am calling it in `main.go` file.
```
package main

import (
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		// Create instance of DynamodbComponentArgs
		dynamodbComponentArgs := DynamodbComponentArgs{
			tableBillingMode:         pulumi.String("PAY_PER_REQUEST"),
			studentsTableName:        pulumi.String("students"),
			classesTableName:         pulumi.String("classes"),
			attendanceTableName:      pulumi.String("attendance"),
			addressTableName:         pulumi.String("addresses"),
			studentParrentsTableName: pulumi.String("student-parents"),
		}
		// Create instance of DynamodbComponent
		dynamodbComponent, err := NewDynamodbComponent(ctx, "dynamodb-component", &dynamodbComponentArgs)
		if err != nil {
			return err
		}
		// Export component outputs
		ctx.Export("studentsTableID", dynamodbComponent.StudentTableID)
		ctx.Export("classesTableID", dynamodbComponent.ClassesTableID)
		ctx.Export("attendanceTableID", dynamodbComponent.AttendanceTableID)
		ctx.Export("addressTableID", dynamodbComponent.AddressTableID)
		ctx.Export("studentParrentsTableID", dynamodbComponent.StudentParrentsTableID)

		return nil
	})
}


```
I hope this helps someone.

**Reply by vilkovtato** [reply]
Hi @mohammadasim, thank you for the example. This would work locally, but what about when the ComponentResource would be in its own repo? This would not work then. After creating a module with ComponentResource and appropriately tagging it, in separated pulumi project I tried to run:
`pulumi package add github.com/xy/some-component-resource@v1.0.0` but getting this error:

error: failed to get schema: could not read plugin [...]: error in compiling Go: go program is not executable, does your program have a 'main' package?

I cannot find proper example for consuming ComponentResources from separate git repo. I believe that consuming ComponentResources from different Pulumi Projects is the main use cases and I'm confused about how to do that.

Would be grateful for any help
